datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,211 @@
+        """
+        Takes in a dict from document id to list of document chunks and inserts them into the index.
+        Return a list of document ids.
+        """
+        # Create a connection
+        conn = psycopg.connect(**self.conn_params)
+        cur = conn.cursor()
+
+        # Initialize a list of ids to return
+        doc_ids: List[str] = []
+
+        # Set synchronous commit
+        cur.execute("SET synchronous_commit = %s" % POSTGRES_SYNCHRONOUS_COMMIT)
+
+        with cur.copy(
+            "COPY %s (doc_id, chunk_id, text, embedding, metadata) FROM STDIN"
+            % POSTGRES_TABLENAME
+        ) as copy:
+            for doc_id, chunk_list in chunks.items():
+                doc_ids.append(doc_id)
+                for chunk in chunk_list:
+                    text = sql.Literal(chunk.text).as_string(conn)
+                    embedding = "[" + ",".join([str(v) for v in chunk.embedding]) + "]"
+
+                    metadata = chunk.metadata.dict()
+                    if "created_at" in list(metadata.keys()):
+                        if metadata["created_at"]:
+                            metadata["created_at"] = to_unix_timestamp(
+                                metadata["created_at"]
+                            )
+                    metadata = json.dumps(metadata)
+
+                    row = (doc_id, chunk.id, text, embedding, metadata)
+                    copy.write_row(row)
+
+        index_statement = (
+            "CREATE INDEX ON %s USING ivfflat (embedding vector_cosine_ops)"
+            % POSTGRES_TABLENAME
+        )
+        cur.execute(index_statement)
+
+        conn.commit()
+        cur.close()
+        conn.close()
+
+        return doc_ids
+
+    async def _query(self, queries: List[QueryWithEmbedding]) -> List[QueryResult]:
+        """
+        Takes in a list of queries with embeddings and filters and
+        returns a list of query results with matching document chunks and scores.
+        """
+        # Prepare query responses and results object
+        results: List[QueryResult] = []
+
+        # Create a connection
+        conn = psycopg.connect(**self.conn_params)
+        cur = conn.cursor()
+
+        for query in queries:
+            query_results: List[DocumentChunkWithScore] = []
+            embedding = ",".join(str(v) for v in query.embedding)
+            query_statement = (
+                "SELECT doc_id, text, metadata, 1 - (%s.embedding <=> ARRAY[%s]::vector) AS score FROM %s ORDER BY score DESC LIMIT %d;"
+                % (
+                    POSTGRES_TABLENAME,
+                    embedding,
+                    POSTGRES_TABLENAME,
+                    query.top_k,
+                )
+            )
+            cur.execute(query_statement)
+            sql_query_results = cur.fetchall()
+
+            query_results: List[DocumentChunkWithScore] = []
+            for result in sql_query_results:
```

Comment:
I'm surprised we can't more transparently pass a result back from psychopg and have to construct something from the intermediate.